yt/frontends/gadget/data_structures.py

```diff
@@ -168,14 +169,40 @@ def validate(self):
         return True
 
 
-class GadgetBinaryFile(ParticleFile):
+class GadgetParticleFile(ParticleFile):
+    # an intermediate abstract class for both gadget types.
+
+    _file_read_mode = None  # either "r" or "rb", set by child class
+    _opening_func = None  # a function handle for opening the file
+
+    @contextlib.contextmanager
+    def transaction(self, handle=None):
+        if handle is None:
+            handle = self._opening_func(self.filename, mode=self._file_read_mode)
+            yield handle
+            handle.close()
+        else:
+            # we have ended up here recursively, so simply yield. The outer
+            # instance will take care of closing the handle.
+            yield handle
+
+
+class GadgetBinaryFile(GadgetParticleFile):
+
+    _file_read_mode = "rb"
+    _opening_func = open
```

Comment:
I am epsilon conflicted about this.  Only epsilon.

Reply:
ya -- I'm not tied to it! the main reason for the new class at all was I was trying to avoid having to define the `.transaction` method in both `GadgetBinaryFile` and `GadgetHDF5File` when they were so similar. but then I needed a way of specifying the function handle to use for opening and ended up with this. But from a readability/understanding perspective, maybe it's simpler to have a little bit of copy/paste here instead of the new class?

Reply:
Nah, this is OK for now.